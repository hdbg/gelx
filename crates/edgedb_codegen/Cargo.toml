[package]
name = "edgedb_codegen"
version.workspace = true
authors.workspace = true
edition.workspace = true
homepage.workspace = true
license.workspace = true
repository.workspace = true
rust-version.workspace = true

[dependencies]
bytes = { workspace = true }
document-features = { workspace = true }
edgedb-derive = { workspace = true }
edgedb-errors = { workspace = true }
edgedb-protocol = { workspace = true, features = ["all-types"] }
edgedb-tokio = { workspace = true, features = ["unstable", "derive"], optional = true }
edgedb_codegen_macros = { workspace = true }
serde = { workspace = true, features = ["derive"], optional = true }
serde_bytes = { workspace = true, optional = true }
typed-builder = { workspace = true }
uuid = { workspace = true }
chrono = { workspace = true, optional = true }
num-bigint = { workspace = true, optional = true }
num-traits = { workspace = true, optional = true }
bigdecimal = { workspace = true, optional = true }

[dev-dependencies]
assert2 = { workspace = true }
edgedb_codegen_core = { workspace = true }
insta = { workspace = true, features = ["ron", "yaml", "redactions"] }
rstest = { workspace = true }
rustversion = { workspace = true }
test-log = { workspace = true, features = ["log", "trace"] }
tokio = { workspace = true, features = ["time", "test-util", "fs"] }
trybuild = { workspace = true }

[lints]
workspace = true

[features]

#! ### Default
#! By default all types are included.
default = ["with_all"]

#! ### Types

## Include the `num-bigint` dependency.
with_bigint = [
	"edgedb_codegen_macros/with_bigint",
	"dep:num-bigint",
	"dep:num-traits",
	"edgedb-protocol/with-num-bigint",
]

## Use the `bigdecimal` crate.
with_bigdecimal = [
	"edgedb_codegen_macros/with_bigdecimal",
	"dep:bigdecimal",
	"dep:num-bigint",
	"dep:num-traits",
	"edgedb-protocol/with-bigdecimal",
]

## Use the `chrono` crate for all dates.
with_chrono = ["edgedb_codegen_macros/with_chrono", "dep:chrono", "edgedb-protocol/with-chrono"]
with_all = [
	"edgedb_codegen_macros/with_all",
	"with_bigint",
	"with_bigdecimal",
	"with_chrono",
	"edgedb-protocol/all-types",
]

## Enable serde for the generated code.
serde = [
	"edgedb_codegen_macros/serde",
	"edgedb-protocol/with-serde",
	"dep:serde",
	"dep:serde_bytes",
	"uuid/serde",
	"chrono/serde",
]

## This turns on the `query` and `transaction` methods and anything that relies on `edgedb-tokio`. The reason to seperate this feature is to enable usage of this macro in browser environments.
query = ["edgedb_codegen_macros/query", "dep:edgedb-tokio"]
