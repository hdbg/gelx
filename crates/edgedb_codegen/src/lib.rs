#![doc = document_features::document_features!()]

pub use edgedb_codegen_macros::*;

pub mod exports {
	pub use ::bytes;
	pub use ::edgedb_derive;
	pub use ::edgedb_errors;
	pub use ::edgedb_protocol;
	#[cfg(feature = "query")]
	pub use ::edgedb_tokio;
	#[cfg(feature = "serde")]
	pub use ::serde;
	#[cfg(feature = "builder")]
	pub use ::typed_builder;
	pub use ::uuid;
}
