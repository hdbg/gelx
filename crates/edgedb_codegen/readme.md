# `edgedb_codegen`

> Generates fully typed rust code from an edgedb schema and inline / file-based queries.

[![Crate][crate-image]][crate-link] [![Docs][docs-image]][docs-link] [![Status][ci-status-image]][ci-status-link] [![Unlicense][unlicense-image]][unlicense-link]

## Installation

To install the `edgedb_codegen` crate you can use the following command.

```bash
cargo add edgedb_codegen
```

Or directly add the following to your `Cargo.toml` file.

```toml
edgedb_codegen = "0.1.0" # replace with the latest version
```

## Description

When working with `edgedb` you often need to write queries and also provide the typed for both the input and output. Your code is only checked at runtime which increases the risk of bugs and errors.

Fortunately, `edgedb` has a query language that is typed and can be converted into types and queried for correctness at compile time.

### Inline Queries

```rust
use edgedb_codegen::edgedb_query;
use edgedb_errors::Result;
use edgedb_tokio::create_client;

// Creates a module called `simple` with a function called `query` and structs
// for the `Input` and `Output`.
edgedb_query!(simple, "select {hello := \"world\", custom := <str>$custom }");

let client = create_client().await?;
let input = simple::Input::builder().custom("custom").build();

// For queries the following code can be used.
let output = simple::query(&client).await?;
```

This macro will generate the following code:

### Query Files

Define a query file in the `queries` directory of your crate called `select_user.edgeql`.

```edgeql
# queries/select_user.edgeql

select User {
  name,
  bio,
  slug,
} filter .slug = <str>$slug;
```

Then use the `edgedb_query` macro to import the query.

```rust
use edgedb_codegen::edgedb_query;
use edgedb_tokio::create_client;


// Creates a module called `select_user` with public functions `transaction` and
// `query` as well as structs for the `Input` and `Output`.
edgedb_query!(select_user);

let client = create_client().await?;
let input = select_user::Input::builder().slug("test").build();

// Generated code can be run inside a transaction.
let result = client.transaction(|mut txn| {
	async move {
		let output = select_user::transaction(&mut txn).await?;
		Ok(output)
	}
}).await?;
```

[crate-image]: https://img.shields.io/crates/v/edgedb_codegen.svg
[crate-link]: https://crates.io/crates/edgedb_codegen
[docs-image]: https://docs.rs/edgedb_codegen/badge.svg
[docs-link]: https://docs.rs/edgedb_codegen/
[ci-status-image]: https://github.com/ifiokjr/edgedb_codegen/workflows/ci/badge.svg
[ci-status-link]: https://github.com/ifiokjr/edgedb_codegen/actions?query=workflow:ci
[unlicense-image]: https://img.shields.io/badge/license-Unlicence-blue.svg
[unlicense-link]: https://opensource.org/license/unlicense
