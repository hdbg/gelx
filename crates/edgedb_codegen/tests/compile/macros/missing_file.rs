use edgedb_codegen::edgedb_query;

fn main() {
	edgedb_query!(example_module);
}
