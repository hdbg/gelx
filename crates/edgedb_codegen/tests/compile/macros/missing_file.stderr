error: failed to read query file at $WORKSPACE/target/tests/trybuild/edgedb_codegen/queries/example_module.edgeql: No such file or directory (os error 2)
 --> tests/compile/macros/missing_file.rs:4:16
  |
4 |     edgedb_query!(example_module);
  |                   ^^^^^^^^^^^^^^
